File: tests/manual/cross_lang_bench/zen/method_call.py
import time

class Toggle:
    def __init__(self, start_state):
        self.state = start_state
    def value(self):
        return self.state
    def activate(self):
        self.state = not self.state
        return self

class NthToggle(Toggle):
    def __init__(self, start_state, max_counter):
        super().__init__(start_state)
        self.count_max = max_counter
        self.count = 0
    def activate(self):
        self.count = self.count + 1
        if self.count >= self.count_max:
            super().activate()
            self.count = 0
        return self

start = time.perf_counter()
n = 100000
val = True
toggle = Toggle(val)

i = 0
while i < n:
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    val = toggle.activate().value()
    i = i + 1

print(toggle.value())

val = True
ntoggle = NthToggle(val, 3)

i = 0
while i < n:
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    val = ntoggle.activate().value()
    i = i + 1

print(ntoggle.value())
print("elapsed:", time.perf_counter() - start)

File: tests/manual/cross_lang_bench/lua/method_call.lua
local Toggle = {}
Toggle.__index = Toggle
function Toggle.new(startState)
    local self = setmetatable({}, Toggle)
    self.state = startState
    return self
end
function Toggle:value() return self.state end
function Toggle:activate()
    self.state = not self.state
    return self
end

local NthToggle = setmetatable({}, { __index = Toggle })
NthToggle.__index = NthToggle
function NthToggle.new(startState, maxCounter)
    local self = setmetatable(Toggle.new(startState), NthToggle)
    self.countMax = maxCounter
    self.count = 0
    return self
end
function NthToggle:activate()
    self.count = self.count + 1
    if self.count >= self.countMax then
        Toggle.activate(self)
        self.count = 0
    end
    return self
end

local start = os.clock()
local n = 100000
local val = true
local toggle = Toggle.new(val)

for i = 1, n do
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
    val = toggle:activate():value()
end

print(toggle:value())

val = true
local ntoggle = NthToggle.new(val, 3)

for i = 1, n do
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
    val = ntoggle:activate():value()
end

print(ntoggle:value())
print("elapsed:", os.clock() - start)
